include/aws/cal/symmetric_cipher.h

```diff
@@ -0,0 +1,263 @@
+ * operations on this cipher again.
+ *
+ * returns AWS_OP_SUCCESS on success. Call aws_last_error() to determine the failure cause if it returns
+ * AWS_OP_ERR;
+ */
+AWS_CAL_API int aws_symmetric_cipher_finalize_encryption(struct aws_symmetric_cipher *cipher, struct aws_byte_buf *out);
+
+/**
+ * Decrypts any remaining data that was reserved for final padding, loads GMACs etc... and if there is any
+ * writes any remaining decrypted data to out. If out is dynamic it will be expanded. If it is not, and
+ * out is not large enough to handle the decrypted output, the call will fail. If you're trying to optimize
+ * to use a stack based array or something, make sure it's at least as large as the size of 2 BLOCKs to account for
+ * padding etc...
+ *
+ * After invoking this function, you MUST call aws_symmetric_cipher_reset() before invoking any encrypt/decrypt
+ * operations on this cipher again.
+ *
+ * returns AWS_OP_SUCCESS on success. Call aws_last_error() to determine the failure cause if it returns
+ * AWS_OP_ERR;
+ */
+AWS_CAL_API int aws_symmetric_cipher_finalize_decryption(struct aws_symmetric_cipher *cipher, struct aws_byte_buf *out);
+
+/**
+ * Resets the cipher state for starting a new encrypt or decrypt operation. Note encrypt/decrypt cannot be mixed on the
+ * same cipher without a call to reset in between them. However, this leaves the key, iv etc... materials setup for
+ * immediate reuse.
+ *
+ * returns AWS_OP_SUCCESS on success. Call aws_last_error() to determine the failure cause if it returns
+ * AWS_OP_ERR;
+ */
+AWS_CAL_API int aws_symmetric_cipher_reset(struct aws_symmetric_cipher *cipher);
+
+/**
+ * Gets the current GMAC tag. If not AES GCM, this function will just copy an empty buffer over.
+ * You would typically call this function after calling aws_symmetric_cipher_finalize_encryption()
+ * to fetch the tag for transmitting it in a cryptographic protocol.
+ *
+ * returns AWS_OP_SUCCESS on success. Call aws_last_error() to determine the failure cause if it returns
+ * AWS_OP_ERR;
+ */
+AWS_CAL_API int aws_symmetric_cipher_get_tag(struct aws_symmetric_cipher *cipher, struct aws_byte_buf *out);
+
+/**
+ * Gets the original intialization vector.
+ *
+ * returns AWS_OP_SUCCESS on success. Call aws_last_error() to determine the failure cause if it returns
+ * AWS_OP_ERR;
+ */
+AWS_CAL_API int aws_symmetric_cipher_get_initialization_vector(
+    struct aws_symmetric_cipher *cipher,
+    struct aws_byte_buf *out);
+
+/**
+ * Gets the original ke.
+ *
+ * returns AWS_OP_SUCCESS on success. Call aws_last_error() to determine the failure cause if it returns
+ * AWS_OP_ERR;
+ */
+AWS_CAL_API int aws_symmetric_cipher_get_key(struct aws_symmetric_cipher *cipher, struct aws_byte_buf *out);
+
+/**
+ * Generates a secure random initialization vector of length len_bytes. If is_counter_mode is set, the final 4 bytes
+ * will be reserved as a counter and initialized to 1 in big-endian byte-order.
```

Comment:
talked offline. I moved this to private. The short answer is CTR and GCM use a 4 byte counter. However, in GCM the 4 byte counter is sort of hidden behind the API because it manages the counter itself. In CTR the counter is exposed in the IV since no hash is associated with it.